File: leetcode/golang/minSessions.go
package golang;

func minSessions(task []int, sessionTime int) (ans int) {
	n := len(tasks)
	m := 1 << n
	// 预处理所有子集的子集和， 复杂度 O(1 + 2 + 4 +  2^(n - 1)) = o(2 ^ n)
	sum := make([]int, m)
	for i, t := range tasks {
		for j, k := 0, 1<<i; j < k; ++j {
			sum[j | k] = sum[j] + t;
		}
	}
	f : make([]int, m)
	for i := range f {
		f[i] = n
	}
	f[0] = 0
	for s := range f {
		// 枚举 s 的所有子集 sub, 若 sub 耗时不超过 sessionTime, 则将 f[s ^ sub] + 1 转移到 f[s] 上
		for sub := s; sub > 0; sub = (sub - 1) & s {
			if sum[sub] <= sessionTime && f[s ^ sub] + 1 < f[s] {
				f[s] = f[s ^ sub] + 1
			}
		}
	}
	return f[m - 1]
}

File: leetcode/cpp/minSessions.cpp
#include <vector>
#include <iostream>
#include <bitset>
#include <climits>

using namespace std;



class Solution {
public:
	int minSessions(vector<int>& tasks, int sessionTime) {
		const int n = tasks.size();
		vector<int> valid(1 << n);
		for (int mask = 1; mask < (1 << n); ++mask) {
			int needTime = 0;
			for (int i = 0; i < n; ++i) {
				if (mask & (1 << i)) {
					needTime += tasks[i];
				}
			}
			if (needTime <= sessionTime) {
				valid[mask] = true;
			}
		}
		vector<int> f(1 << n, INT_MAX / 2);  
		f[0] = 0;
		for (int mask = 1; mask < (1 << n); ++mask) {
			cout << "mask: " << std::bitset<10>(mask) << endl;
			for (int subset = mask; subset; subset = (subset - 1) & mask) {
				cout << " " << std::bitset<10>(subset);
				if (valid[subset]) {
					f[mask] = min(f[mask], f[mask ^ subset] + 1);
				}
			}
			cout << endl;
		}
		return f[(1 << n) - 1];
	}
};


int main() {
	vector<int> tasks = {2, 3, 3, 4, 4, 4, 5, 6, 7, 10};
	int sessionTime = 12;

	Solution s;
	cout << "s: " << s.minSessions(tasks, sessionTime) << endl;


}












